doc/designs/evp_skey_multi.md

```diff
@@ -0,0 +1,67 @@
+Simultaneous derivation of several EVP_SKEY objects
+===================================================
+
+There are situations where we need to derive several symmetric keys
+simultaneously.  The most relevant one for OpenSSL is TLS protocol, when we
+need to derive 2-4 keys, depending on the protocol version. With raw bytes
+buffer, the approach was to derive a combined buffer of the necessary length
+and chop it. It doesn't work this way for EVP_SKEY objects.
+
+This document proposes API and a general approach to deal with such situations.
+
+Proposed API to derive an EVP_SKEY objects
+------------------------------------------
+
+As all the objects are derived in one transaction, we can store a single opaque
+pointer keeping all the keys inside in the EVP_KDF_CTX object, and provide the
+API for access to a particular object.
+
+However, this will require extending the API on the libcrypto/libssl side and
+complications on a provider side.
+
+Libcrypto API
+-------------
+
+The initialization of EVP_KDF_CTX for simultaneous derivation is done via
+call to `EVP_KDF_CTX_set_params`.
+
+To derive the opaque keys and and bytes buffers, we use the function
+
+```C
+int *EVP_KDF_derive_SKEYs(EVP_KDF_CTX *ctx, EVP_SKEYMGMT *mgmt,
+                          const char **key_types, const size_t *keylengths, size_t keynum,
+                          char **ivptrs, const size_t *ivlengths, size_t ivnum,
+                          const char *propquery, const OSSL_PARAM params[]);
```

Comment:
This function now returns an array of `int`.  This design should probably say what those `int`s are

Reply:
Considering the provider API further down, is `int *` simply an error, and the return type should simply be `int`?